Fix regression in 0.49RC1 for first-class functions

pearu opened the pull request:
Fixes #5470 

<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
mailing list https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

stuartarchibald commented:
Thanks for the patch. Should dc1c088 be present in the patchset ?

pearu commented:
No, it should not.

esc commented:
@pearu would it be possible to rebase to remove the commit: https://github.com/numba/numba/commit/dc1c088654e9a4073e7a26b1113599077f92baa2

pearu commented:
> @pearu would it be possible to rebase to remove the commit: [dc1c088](https://github.com/numba/numba/commit/dc1c088654e9a4073e7a26b1113599077f92baa2)

Done.

stuartarchibald commented:
/AzurePipelines run

stuartarchibald commented on numba/core/utils.py:
```diff
@@ -472,7 +472,7 @@ def unify_function_types(numba_types):
 
 
-def unified_function_type(numba_types):
+def unified_function_type(numba_types, require_precise=True):
```
This kwarg addition should be documented. What's it's purpose and the effect of having it not set?

pearu replied:
Updated unified_function_type docstring.

stuartarchibald commented:
Thanks for the patch, it's queued for review.

esc commented:
Note for reviewers: there was some back and forth on #5470 - but it does appear like this PR does indeed fix that issue. Thank you @pearu !

stuartarchibald commented on numba/core/utils.py:
```diff
@@ -472,11 +472,31 @@ def unify_function_types(numba_types):
-    If any of the Numba function types is a Dispatcher type, the
-    unified function type will be UndefinedFunctionType instance.
+def unified_function_type(numba_types, require_precise=True):
+    """Return unified Numba function type if possible.
```
```suggestion
    """Returns a unified Numba function type if possible.
```

pearu replied:
Fixed.

stuartarchibald commented on numba/core/utils.py:
```diff
@@ -472,11 +472,31 @@ def unify_function_types(numba_types):
+      least two different Numba function type instances.
+
+    If ``numba_types`` contains a Numba dispatcher type, the unified
+    Numba function type will be unprecise ``UndefinedFunctionType``
```
```suggestion
    Numba function type will be an imprecise ``UndefinedFunctionType``
```

pearu replied:
Fixed.

stuartarchibald commented on numba/core/utils.py:
```diff
@@ -472,11 +472,31 @@ def unify_function_types(numba_types):
+    Numba function type will be unprecise ``UndefinedFunctionType``
+    instance, or None when ``require_precise=True`` is specified.
+
+    Specifying ``require_precise=False`` enables unifying unprecise
```
```suggestion
    Specifying ``require_precise=False`` enables unifying imprecise
```

pearu replied:
Fixed.

stuartarchibald commented on numba/tests/test_function_type.py:
```diff
@@ -187,7 +187,7 @@ def foo(f, g):
         sig = int64(int64)
 
-        for decor in [mk_cfunc_func(sig), mk_wap_func(sig), njit_func,
+        for decor in [mk_cfunc_func(sig), mk_wap_func(sig), # njit_func,
```
If this is now unsupported, consider removal instead of commenting it out? (same comment applies throughout).

pearu replied:
The tests are now cleaned up.

stuartarchibald commented on numba/tests/test_function_type.py:
```diff
@@ -917,14 +917,15 @@ def foo_template(funcs, x):
-        self.assertEqual(r, 3)
-        # the Tuple type of foo first argument is UndefinedFunctionType:
-        self.assertEqual(foo.signatures[0][0].dtype.is_precise(), False)
+        if 0:
+            # Problem:
+            a = jit(nopython=True)(lambda x: x + 1)
+            b = jit(nopython=True)(lambda x: x + 2)
+            foo = jit(nopython=True)(foo_template)
+            r = foo((a, b), 0)
+            self.assertEqual(r, 3)
+            # the Tuple type of foo first argument is UndefinedFunctionType:
+            self.assertEqual(foo.signatures[0][0].dtype.is_precise(), False)
```
Perhaps comment this block along with some explanation of why it's not going to work opposed to `if 0:` ?

pearu replied:
I have updated the statement of the problem that is about interpreting certain dispatcher types as precise (compilation is disabled and there exists exactly one compiled overload).
The code in `if 0:` will not work because of the disabled support of using dispatcher instances as tuple items. So, removed `if 0:` block.

pearu commented on numba/core/utils.py:
```diff
@@ -472,11 +472,31 @@ def unify_function_types(numba_types):
+    ----------
+    numba_types : tuple
+      Numba type instances.
+    require_precise : boo
```
Fixed.

pearu commented:
Thanks, @stuartarchibald for the review. I have cleaned up the tests and updated the docs.

Using jit-decorated first-class functions in tuples has now a restriction that at least of the items must have a precise function type. So the support for the tuple of dispatcher instances is not completely disabled. For the full support, some further work is required on the numba type-inference algorithm.

sklam commented on numba/tests/test_function_type.py:
```diff
@@ -1015,3 +998,33 @@ def foo_bad2good():
+        formulae_foo = (foo1, foo1)
+
+        @njit()
+        def bar_scalar(_, __):
```
Can we make this function uses `foo1` and `foo2` so that we know that the compiled function is correct?

stuartarchibald commented on docs/source/reference/types.rst:
```diff
@@ -162,12 +162,16 @@ Context                    jit compiled   cfunc compiled   WAP objects
 Can be used as arguments   yes            yes              yes
 Can be called              yes            yes              yes
-Can be used as items       yes            yes              yes
+Can be used as items       yes1           yes              yes
```
Would `yes*` be a more common thing to write? and then below have `* at least one of the...` .

stuartarchibald commented on docs/source/reference/types.rst:
```diff
@@ -162,12 +162,16 @@ Context                    jit compiled   cfunc compiled   WAP objects
 Automatic overload         yes            no               no
 ========================   ============   ==============   ===========
 
+Here ``yes1`` means that at least one of the items in a sequence of
+first-class function objects has a precise type.
```
Should this be a clarification of the restriction, e.g.
> at least one of the items in the sequence of first-class function objects must have a precise type.

?

stuartarchibald commented on numba/tests/test_function_type.py:
```diff
@@ -899,43 +888,36 @@ def foo(disp1, disp2, sel):
+        # In general, the type of a dispatcher instance is imprecise
+        # and when used as an input to type-inference, the typing will
+        # likely fail. However, if a dispatcher instance contains
+        # exactly one overload and the compilation the dispatcher is
```
```suggestion
        # exactly one overload and the compilation for dispatcher is
```

stuartarchibald commented on numba/tests/test_function_type.py:
```diff
@@ -899,43 +888,36 @@ def foo(disp1, disp2, sel):
         r = foo((a, b), 0)
         self.assertEqual(r, 3)
-        # the Tuple type of foo first argument is FunctionType:
+        # the Tuple type of foo first argument is precise FunctionType:
```
```suggestion
        # the Tuple type of foo's first argument is a precise FunctionType:
```

stuartarchibald commented:
Thanks for doing all the work on this @pearu, I branched off this in #5529 taking all the history you created and added in a couple of patches to address the remaining review comments. Will close this but the credit for fixing this is most definitely yours! Thanks again.

pearu commented:
Thanks, @stuartarchibald ! Sorry for the delay in my part to address these issues.

stuartarchibald commented:
@pearu no problem, and no delay at all, @sklam and I just had some external constraints applied on getting 0.49.0RC2 out so we decided to just move stuff as fast as we could!
